pkg/controller/core.oam.dev/v1alpha2/application/dispatch/dispatch.go

```diff
@@ -0,0 +1,166 @@
+/*
+Copyright 2021 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dispatch
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+	"github.com/oam-dev/kubevela/pkg/utils/apply"
+)
+
+// NewAppManifestsDispatcher creates an AppManifestsDispatcher.
+// A K8s client and an ApplicationRevision are required while the resource tracker is optional.
+// If the resource tracker is not provided, the dispatcher will skip GC automatically.
+func NewAppManifestsDispatcher(c client.Client, appRev *v1beta1.ApplicationRevision, latestRT *v1beta1.ResourceTracker) *AppManifestsDispatcher {
+	return &AppManifestsDispatcher{
+		c:          c,
+		applicator: apply.NewAPIApplicator(c),
+		appRev:     appRev.DeepCopy(),
+		oldRT:      latestRT.DeepCopy(),
+		gcHandler:  NewGCHandler(c, appRev.Namespace),
+	}
+}
+
+// AppManifestsDispatcher dispatches application manifests into K8s
+type AppManifestsDispatcher struct {
+	c          client.Client
+	applicator apply.Applicator
+
+	appRev    *v1beta1.ApplicationRevision
+	oldRT     *v1beta1.ResourceTracker
+	skipGC    bool
+	gcHandler GarbageCollector
+
+	appRevName string
+	namespace  string
+	newRTName  string
+	newRT      *v1beta1.ResourceTracker
+}
+
+// SkipGarbageCollect makes dispatcher to skip GC after applying resources
+// It's helpful in a rollout scenario where new revision is going to create a new workload while the old one should not
+// be deleted before rollout is termined.
+func (a *AppManifestsDispatcher) SkipGarbageCollect() {
+	a.skipGC = true
+}
+
+// Dispatch apply manifests into k8s and return a resource tracker recording applied resources
+// If the latest resource tracker is offered to dispatcher, it will do GC after applying.
+func (a *AppManifestsDispatcher) Dispatch(ctx context.Context, manifests []*unstructured.Unstructured) (*v1beta1.ResourceTracker, error) {
+	a.complete()
+	if err := a.createResourceTracker(ctx, manifests); err != nil {
+		return nil, err
+	}
+	if err := a.applyManifests(ctx, manifests); err != nil {
+		return nil, err
+	}
+	if !a.skipGC && a.oldRT != nil && a.oldRT.Name != a.newRTName {
```

Comment:
Can a.oldRT be nil? I found  Line 47 in `NewAppManifestsDispatcher` has init the value with `DeepCopy`

Reply:
Yes, it could be nil. Code is updated and I use another way to pass in old rt.